runtime/intern.c

```diff
@@ -110,19 +116,32 @@ static struct caml_intern_state* get_intern_state (void)
   s->intern_input = NULL;
   s->obj_counter = 0;
   s->intern_obj_table = NULL;
-  s->intern_stack = s->intern_stack_init;
-  s->intern_stack_limit = s->intern_stack + INTERN_STACK_INIT_SIZE;
   s->intern_dest = NULL;
+  init_intern_stack(s);
 
   Caml_state->intern_state = s;
   return s;
 }
 
+static struct caml_intern_state* get_intern_state (void)
+{
+  Caml_check_caml_state();
+
+  if (Caml_state->intern_state == NULL)
+    caml_fatal_error (
+      "intern_state not initialized: "
+      "this function can only be called from a `caml_input_*` entrypoint."
```

Comment:
I'm ok with the first line of the error message "intern_state not initialised". I don't know whether the rest of the message is clarifying the error more "this function can only be called from a `caml_input_*` entrypoint.". 

When will this error occur? Is it that `caml_deserialise_*` was called directly without calling one of the `caml_input_*` functions?

Reply:
> When will this error occur? Is it that `caml_deserialise_*` was called directly without calling one of the `caml_input_*` functions?

Exactly. In that case, before this PR, the field `intern_src` pointing to the source of the marshalled data won't have been populated, and would be accessed, leading to the nullptr dereference. By splitting initialization of the structure (in `init_intern_state`) and then access (with `get_intern_state`), we can check whether it has been provided a marshalled data source, and error if not (admittedly nicer than a segfault).

Reply:
May I request that you add a testcase for this under `testsuite/tests/lib-marshal`? 

Also, the message can be more clear. Instead of

> "this function can only be called from a `caml_input_*` entrypoint."

How about

> "It is likely that caml_deserialize_* function was called without going through `caml_input_*`.

Reply:
> May I request that you add a testcase for this under testsuite/tests/lib-marshal?

Maybe easier said than done? As the functions raise a fatal error, I can only test one of them by process.

Reply:
> Maybe easier said than done?

This is doable. See `.run` files in the test suite for how it is done. But it is not worth the trouble here as the bytecode and native code behaviours will diverge. Bytecode will perform an intern at 

https://github.com/ocaml/ocaml/blob/225f2f3aebb0540582e61c2ab2b23c57f71d7c73/runtime/startup_byt.c#L566

before user code is run. This will cause any test to not cause a fatal error, which is different from native code.